lib/auth/access_request_test.go

```diff
@@ -0,0 +1,379 @@
+			// encoded in the cert
+			checkCerts(t,
+				certs,
+				tc.expectRoles,
+				[]string{"root"},
+				[]string{req.GetName()},
+				tc.requestResources)
+
+			elevatedCert, err := tls.X509KeyPair(certs.TLS, privKey)
+			require.NoError(t, err)
+			elevatedClient := server.NewClientWithCert(elevatedCert)
+
+			// renew elevated certs
+			newCerts, err := elevatedClient.GenerateUserCerts(ctx, proto.UserCertsRequest{
+				PublicKey: pubKey,
+				Username:  tc.requester,
+				Expires:   time.Now().Add(time.Hour).UTC(),
+				// no new access requests
+				AccessRequests: nil,
+			})
+			require.NoError(t, err)
+
+			// inspite of providing no access requests, we still have elevated
+			// roles and the certicate shows the original access request
+			checkCerts(t,
+				newCerts,
+				tc.expectRoles,
+				[]string{"root"},
+				[]string{req.GetName()},
+				tc.requestResources)
+
+			// attempt to apply request in DENIED state (should fail)
+			require.NoError(t, server.Auth().SetAccessRequestState(ctx, types.AccessRequestUpdate{
+				RequestID: req.GetName(),
+				State:     types.RequestState_DENIED,
+			}))
+			_, err = generateCerts(req.GetName())
+			require.ErrorIs(t, err, trace.AccessDenied("access request %q has been denied", req.GetName()))
+
+			// ensure that once in the DENIED state, a request cannot be set back to PENDING state.
+			require.Error(t, server.Auth().SetAccessRequestState(ctx, types.AccessRequestUpdate{
+				RequestID: req.GetName(),
+				State:     types.RequestState_PENDING,
+			}))
+
+			// ensure that once in the DENIED state, a request cannot be set back to APPROVED state.
+			require.Error(t, server.Auth().SetAccessRequestState(ctx, types.AccessRequestUpdate{
+				RequestID: req.GetName(),
+				State:     types.RequestState_APPROVED,
+			}))
+
+			// ensure that identities with requests in the DENIED state can't reissue new certs.
+			_, err = elevatedClient.GenerateUserCerts(ctx, proto.UserCertsRequest{
+				PublicKey: pubKey,
+				Username:  tc.requester,
+				Expires:   time.Now().Add(time.Hour).UTC(),
+				// no new access requests
+				AccessRequests: nil,
+			})
+			require.ErrorIs(t, err, trace.AccessDenied("access request %q has been denied", req.GetName()))
+		})
+	}
+}
+
+// checkCerts checks that the ssh and tls certs include the given roles, logins,
+// accessRequests, and resourceIDs
+func checkCerts(t *testing.T,
+	certs *proto.Certs,
+	roles []string,
+	logins []string,
+	accessRequests []string,
+	resourceIDs []types.ResourceID,
+) {
+	sshCert, err := sshutils.ParseCertificate(certs.SSH)
+	require.NoError(t, err)
+
+	tlsCert, err := tlsca.ParseCertificatePEM(certs.TLS)
+	require.NoError(t, err)
+	tlsIdentity, err := tlsca.FromSubject(tlsCert.Subject, tlsCert.NotAfter)
+	require.NoError(t, err)
+
+	rawSSHCertRoles := sshCert.Permissions.Extensions[teleport.CertExtensionTeleportRoles]
+	sshCertRoles, err := services.UnmarshalCertRoles(rawSSHCertRoles)
+	require.NoError(t, err)
+	sort.Strings(roles)
```

Comment:
Instead of sorting and using `require.Equal`, you can probably get away with `require.ElementsMatch()`.

Reply:
Thank you! I was hoping `require.ElementsMatch()` existed and went looking for it, but didn't know the name and couldn't find it. This is much better.